### PR TITLE
[fem] Move SchurComplement out of dev

### PR DESCRIPTION
**Pull request by xuchenhan-tri**
Towards #16492

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498)

### CODE REVIEW

**Review by xuchenhan-tri — COMMENTED** [review]
+@joemasterjohn for feature review please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498#-:-MuwxMiC0q409m4gt-Fx:bqwin3e)* status: LGTM missing from assignee joemasterjohn, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @joemasterjohn)

**Review by joemasterjohn — COMMENTED** [review]
:lgtm:. +@rpoyner-tri for platform review, please.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498#-:-MvPAmHJDZFNEHVAqol0:becwctl)* status: 5 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform) (waiting on @rpoyner-tri and @xuchenhan-tri)

___
*[multibody/fem/schur_complement.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP3oT-05n7Df1-Orj0:-MvP3oT-05n7Df1-Orj1:b-exefma) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.h#L19)):*
> ```objc
>      Bᵀx + Dy =  0     (2)
>  where M = [A B; Bᵀ D], zᵀ = [xᵀ yᵀ], cᵀ = [aᵀ 0ᵀ], and A(size p-by-p),
>  D(size q-by-q) and M(size p+q-by-p+q) are positive definite, we may choose to
> ```

nit: Just a nitpick about the language here. "we may choose" might imply something about our intentions in code. "One can solve the system using Schur complement." might be a more agnostic way to state what you're saying.
___
*[multibody/fem/schur_complement.h, line 61 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP6qkK0cVcDgZlW8nH:-MvP6qkK0cVcDgZlW8nI:b21ibti) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.h#L61)):*
> ```objc
> 
>   /* Solves for y given the solution for x assuming the right hand side takes
>   the form  [aᵀ 0ᵀ]x using the fact that y = -D⁻¹Bᵀx. See class documentation.
> ```

nit: Is `[aᵀ 0ᵀ]x` a typo and should be `[aᵀ 0ᵀ]`? I'm having trouble finding which rhs this is talking about.
___
*[multibody/fem/schur_complement.cc, line 16 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP7ERG6CF-2lSIOck_:-MvP7ERG6CF-2lSIOcka:beiwvup) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.cc#L16)):*
> ```cpp
>     const Eigen::Ref<const Eigen::SparseMatrix<T>>& D)
>     : p_(A.rows()), q_(D.rows()) {
>   DRAKE_DEMAND(A.cols() == p_);
> ```

nit: If you rephrase this as `DRAKE_DEMAND(A.cols() == A.rows())` etc. the error message might come out a bit nicer when triggered.
___
*[multibody/fem/schur_complement.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP7d3A0OVitfvM8-5K:-MvP7d3A0OVitfvM8-5L:bkhh9tn) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.cc#L29)):*
> ```cpp
>     // D_complement is empty, same as default.
>   } else {
>     /* Note that since D is SPD, we could use a sparse Cholesky factorization
> ```

nit: Is it worth adding a `TODO(): Change to sparse Cholesky factorization.`, or is the license restriction likely to never go away?
___
*[multibody/fem/test/schur_complement_test.cc, line 48 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvPAXDq8Cu_Koimf0zQ:-MvPAXDq8Cu_Koimf0zR:bsrzz7u) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/test/schur_complement_test.cc#L48)):*
> ```cpp
>   EXPECT_TRUE(CompareMatrices(s.get_D_complement(), expected_D_complement,
>                               std::numeric_limits<double>::epsilon()));
> ```

The tests are juts missing coverage for the prescribed constructor (kcov complained). Can you just add a few lines to the analytic test that exercise it, please?



_Code quote (from multibody/fem/schur_complement.h):_
```objc
SchurComplement(MatrixX<T>&& D_complement, MatrixX<T>&& neg_Dinv_B_transpose);
```

**Review by rpoyner-tri — COMMENTED** [review]
hm, conflicts with master now -- needs rebase. Looks generally good. I'll return when the conflict is fixed.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498#-:-MvPEU1W5rGy9dGCAkLX:bdk22vp)* status: 5 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform) (waiting on @xuchenhan-tri)

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498)* status: 3 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform) (waiting on @joemasterjohn and @xuchenhan-tri)

___
*[multibody/fem/schur_complement.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP3oT-05n7Df1-Orj0:-MvQ1qG-AwWgWt5lapZl:b-xwwb0a) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.h#L19)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Done, thanks.
</blockquote></details>

I'm not actually seeing a change in `r2`?
___
*[multibody/fem/schur_complement.h, line 61 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP6qkK0cVcDgZlW8nH:-MvQ1v7z33KMPPYJ6DNv:b-ruiapl) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.h#L61)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I reworded to cut down on irrelevant information,
</blockquote></details>

ditto -- missing a push?
___
*[multibody/fem/test/schur_complement_test.cc, line 48 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvPAXDq8Cu_Koimf0zQ:-MvQ2oVmBHBE67YX25SU:b-c3vx4p) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/test/schur_complement_test.cc#L48)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Done, thanks!
</blockquote></details>

missing a push?

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498)* status: 3 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform) (waiting on @joemasterjohn and @xuchenhan-tri)

___
*[multibody/fem/schur_complement.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP3oT-05n7Df1-Orj0:-MvQ7fO73CKY4YG7ixN_:bcu37j7) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.h#L19)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

I'm not actually seeing a change in `r2`?
</blockquote></details>

Oops, yeah forgot to push. `r3` is the revision to be reviewed.

**Review by joemasterjohn — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498)* status: LGTM missing from assignee rpoyner-tri(platform) (waiting on @xuchenhan-tri)

___
*[multibody/fem/schur_complement.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16498#-MvP7d3A0OVitfvM8-5K:-MvQ9gnxEV2PpWjEaF8_:b-k6rpao) ([raw file](https://github.com/robotlocomotion/drake/blob/eed4096b1c615aea6630aa358a02c38ded2d7cab/multibody/fem/schur_complement.cc#L29)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Unfortunately, I don't see a reason for the license restriction to go away :(
On the bright side, we are moving away from Eigen::SparseMatrix, and we can retire this method all together when that happens.
</blockquote></details>

Sounds good, thanks!

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498)* status: LGTM missing from assignee rpoyner-tri(platform) (waiting on @xuchenhan-tri)

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16498#-:-MvQCHEa7P1HUu0_hP6X:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),joemasterjohn (waiting on @xuchenhan-tri)
